tsdb/engine/tsm1/compact.go

```diff
@@ -1128,33 +1128,34 @@ func (c *Compactor) write(path string, iter KeyIterator, throttle bool, logger *
 	// in memory.
 	if iter.EstimatedIndexSize() > 64*1024*1024 {
 		w, err = NewTSMWriterWithDiskBuffer(limitWriter)
-		if err != nil {
-			return err
-		}
 	} else {
 		w, err = NewTSMWriter(limitWriter)
-		if err != nil {
-			return err
-		}
 	}
-
+	if err != nil {
+		// Close the file and return if we can't create the TSMWriter
+		return errors.Join(err, fd.Close())
+	}
 	defer func() {
+		errs := make([]error, 0, 2)
+		errs = append(errs, err)
 		closeErr := w.Close()
 		if err == nil {
+			// Save closeErr as err for later checks
 			err = closeErr
 		}
```

Comment:
On the whole, the handling of `err` and `errs` through the defer are confusing.

It looks like every path out of the defer will ends with `err = errors.Join(errs...)`. If this is true, does shuffling `closeErr` into `err` accomplish anything?

Reply:
I'm now rewriting the `defer` function, because, as you say, it's a mess.  It's too permissive (the `OpenFile` call failure should be tested for `os.ErrExist` as the only actually compaction in progress indicator, and other errors should cause an abort.  Similarly, a failure on `Close` should abort the `ErrMaxBlocksExceeded` and `errMaxFileExceeded` rolling over into a new file.).  The root confusion is treating exceeding the limits as an error, when it is really a separate state.  Errors cause the `tmp` file's removal; non-errors do not.  Exceeding the limits causes the creation of another `tmp` file.